### PR TITLE
Feat: #210 support pbn post before commenting open

### PR DESCRIPTION
**Pull request by ianliuwk1019**
For supporting Public Notices open for public before commenting open date ticket #210 .
- Implement flow (# - 2) for this enhancement for simple architecture we discussed previously with just one new `post_date` column in public_notice entity. (frontend labeled as 'Notice Publishing Date' suggested by Julius.)
- Allow user to select 'Notice Publishing Date' date (YYYY-MM-DD) from admin - public notice (add/edit) before PUBLISH pushed.
- New 'Notice Publishing Date' field empty is allowed.
- Popup (suggested by Julius) warning for datepicker when commenting start date has not bee set, so no selection for 'Notice Publishing Date'.
- Migration script for setting new post_date column to commenting_open_date if not set (but exclude INITIAL state).
- During PUBLISH is pushed, validate post_date value is correct if present and if not present, set the value to commenting_open_date.
- New api-client is generated.
- Unit tests added for validation.

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-393.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-393.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-393.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

File: api/src/app/modules/project/project.service.ts [hunk 1]
```diff
@@ -307,6 +309,7 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
 
     const options = {relations: []};
     options.relations.push('submissions'); // add this extra relation for later use.
+		options.relations.push('publicNotices');
```
Comment (basilv) [inline api/src/app/modules/project/project.service.ts:311]: why are we doing this?

File: api/src/app/modules/project/project.service.ts [hunk 2]
```diff
@@ -503,6 +507,27 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
...
+        const postDate = publicNotices[0].postDate;
+        const commentingOpenDate = entity.commentingOpenDate;
+        if (postDate && !isPNPostdateOnOrBeforeCommentingOpenDate(postDate, commentingOpenDate)) {
+			throw new BadRequestException(`Unable to transition FOM ${entity.id} to ${stateTransition}. 
+            Online Public Notice post date ${postDate} should be on or before commenting start date 
+            ${commentingOpenDate}.`);
+        }
+
+        // Must be at least one day after publish is pushed
+        const dayDiff = DateTimeUtil.diffNow(postDate, DateTimeUtil.TIMEZONE_VANCOUVER, 'day');
```
Comment (basilv) [inline api/src/app/modules/project/project.service.ts]: What if postDate is null here?

File: api/src/app/modules/project/project.service.ts [hunk 3]
```diff
@@ -725,5 +750,53 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
...
+				const pnId = publicNotices[0].id;
+				const commentingOpenDate = projectEntity.commentingOpenDate;
+				const updateFields = 
+				{
+					postDate: commentingOpenDate,
+					updateUser: 'system',
+					updateTimestamp: new Date(),
+					revisionCount: () => "revision_count + 1"
+				}
+				const updatedCount = (await this.repository
```
Comment (basilv) [inline api/src/app/modules/project/project.service.ts]: Why are we doing this? (logic with updatedCount - I'm not following the reason)

File: api/src/app/modules/project/project.service.ts [hunk 4]
```diff
@@ -725,5 +750,53 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
...
+/**
+ * Public Notice postDate validation: on or before commenting start date.
+ * This is exported validation to be conveniently used in other service.
+ * @param postDate YYYY-MM-DD date string for public notice post date.
+ * @param commentingOpenDate YYYY-MM-DD date string for FOM (Project) commentingOpenDate
+ * @returns true only if postDate on or before commentingOpenDate; 
+ */
+export function isPNPostdateOnOrBeforeCommentingOpenDate(postDate: string, commentingOpenDate: string) {
+	if (_.isEmpty(commentingOpenDate) || !isValidDateOnlyString(commentingOpenDate)) {
+			throw new InternalServerErrorException(`Invalid argument commentingOpenDate: ${commentingOpenDate}`)
```
Comment (basilv) [inline api/src/app/modules/project/project.service.ts]: I think this would be a bad request exception instead (same with below), since both fields are coming from the front-end, but its debateable.

File: api/src/app/modules/project/public-notice.service.ts [hunk 5]
```diff
@@ -177,8 +187,27 @@ export class PublicNoticeService extends DataService<PublicNotice, Repository<Pu
     response.revisionCount = entity.revisionCount;
     response.operationStartYear = entity.operationStartYear;
     response.operationEndYear = entity.operationEndYear;
+		response.postDate = entity.postDate;
     return response;
   }
 
-}
-
+	// Override: for some dto property business values transformation.
```
Comment (basilv) [inline api/src/app/modules/project/public-notice.service.ts]: Why is this needed?

File: api/src/migrations/main/1687301343450-public-notice-post-date.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    
+        -- add new column - post_date, nullable (for existing data before new column exists)
+            ALTER TABLE app_fom.public_notice ADD COLUMN post_date date;
+
+        -- comment on column - post_date
+            COMMENT ON COLUMN app_fom.public_notice.post_date IS
+            'Date planed for online public notice posted (visible to public) before the FOM reaches the commenting open period';
+
+        -- set post_date value to app_fom.project.commenting_open_date for the past data as default value.
+            update app_fom.public_notice pn set post_date = (
```
Comment (basilv) [inline api/src/migrations/main/1687301343450-public-notice-post-date.js:26]: so for existing public notices, set the post date to commenting open. Maybe we should exclude projects in state INITIAL?

**Review by basilv — CHANGES_REQUESTED** [review]
Some changes suggested. I somehow didn't see the changes for retrieving more public notices and the caching, I'll still ideally want to take a look at that. (I assume those changes are made, but I just can't easily find them in GitHub's PR review )

File: api/src/app/modules/project/project.service.ts (hunk 2 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts]: **Good point**, but strange things is I ran flow from "admin" ui and still works fine when postDate is null and final outcome seem to be correct. Strange, somewhere if it is null probably already skipped this part of validation. I will check why I didn't catch this. Thanks.

File: api/src/app/modules/project/public-notice.service.ts (hunk 5 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/public-notice.service.ts]: Sorry, the comment line is confusing. I should remove that.
But now I know why I override convertDto() method from dataService. Not sure if you have time for discussion (for design decision).

File: api/src/app/modules/project/project.service.ts (hunk 2 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts]: The intention initially was, if postDate is empty, no need to do validation (since it will be set to default commenting open date when FOM state changes to PUBLISH).

Ha, I checked why the flow was still good, turns out even if date is null, the datejs library returns a null for the "diff". So, accidently, the next line is still evaluated to be false so the validation pass, leading that I didn't detect this bug when running different scenarios using frontend flow. Anyway, I added a condition in. Thanks for catching this.

File: api/src/migrations/main/1687301343450-public-notice-post-date.js (hunk 6 above)
Comment (ianliuwk1019) [reply api/src/migrations/main/1687301343450-public-notice-post-date.js:26]: Hi Basil, I have added this exclusion in script. However, github section on your review seem a bit strange and does not reflect the change. Please go to the file to verify again for this script please. Thank you Basil.

File: api/src/app/modules/project/project.service.ts (hunk 1 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts:311]: Interesting... not sure why on the pr the line does not align but on vscode it is....

During the workflow change request (when 'PUBLISH' is pushed/requested), it needs to add one more relation ('publicNotices') to load from TypeORM (like previously the 'submission' relation) so it can validate the public_notice.post_date later. If this is not added, relation will be null (even though database exists the relation).
We could load is separately later during validation but loading it initially should make more sense.
image

**Comment by DerekRoberts** [comment]
@ianliuwk1019 @basilv GitHub has been acting up and not updating PRs. It's infrequent, but if you see this, please do not merge the PR until it catches up.

File: api/src/app/modules/project/project.service.ts (hunk 3 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts]: At the project service, during the workflow_state_change to "PUBLISH", this is where the code update the publicNotice entity post_date to set to commenting_open_date if post_date is empty.

File: api/src/app/modules/project/project.service.ts (hunk 1 above)
Comment (basilv) [reply api/src/app/modules/project/project.service.ts:311]: This is fine.

File: api/src/app/modules/project/public-notice.service.ts (hunk 5 above)
Comment (basilv) [reply api/src/app/modules/project/public-notice.service.ts]: Discussed, This is okay, it would be slightly better I feel to do this in the public-notice.service, but data.service isn't really designed for this validation scenario either.

File: api/src/app/modules/project/project.service.ts (hunk 4 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts]: Since this is refactored into DateTimeUtil class, I also change it to throw just Error (we don't currently has custom Error on our end), hope that will make more sense. Currently this is like an util function and does not directly involved in controller returns, could be called from different places, so I gives it an general error. But I would be happy to change it to BadRequestException if you prefer.

File: api/src/migrations/main/1687301343450-public-notice-post-date.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+                select *
+                from app_fom.project
+                where workflow_state_code != 'INITIAL' 
+            )
+            update app_fom.public_notice pn set post_date = (
+                select p.commenting_open_date from project_for_update p 
+                where p.project_id = pn.project_id )
+            from project_for_update pu
+            where post_date is null and pn.project_id = pu.project_id;
+        `);
```
Comment (ianliuwk1019) [inline api/src/migrations/main/1687301343450-public-notice-post-date.js:31]: Hi @basilv, 
I have adjusted data migration script for "public_notice.post_date", I did test it on PgAdmin, but it is worth for you to take another look! Thank you so much.

File: api/src/app/modules/interaction/interaction.dto.ts [hunk 8]
```diff
@@ -16,24 +15,24 @@ export function IsISODateOnlyString(validationOptions?: ValidationOptions) {
...
-Note: "2023-06-07" and "06-07-2023" and "2023/06/07" etc are treated valid from dayjs library and all
-    are translated into "2023-06-07" by the dayjs library.
-For now it does not provide flexbility to check other format.
-*/
-export function isValidDateOnlyString(value: string): boolean {
-    dayjs.extend(customParseFormat);
-    return !_.isEmpty(value) && value.length == 10 && dayjs(value, DateTimeUtil.DATE_FORMAT, true).isValid();
-}
+// /*
+// Limited to only check date string value with backend defined date format: DateTimeUtil.DATE_FORMAT.
```
Comment (basilv) [inline api/src/app/modules/interaction/interaction.dto.ts]: remove this commented out code?

File: api/src/app/modules/project/project.service.ts [hunk 9]
```diff
@@ -503,6 +506,28 @@ export class ProjectService extends DataService<Project, Repository<Project>, Pr
...
+					if (postDate && !DateTimeUtil.isPNPostdateOnOrBeforeCommentingOpenDate(postDate, commentingOpenDate)) {
+						throw new BadRequestException(`Unable to transition FOM ${entity.id} to ${stateTransition}. 
+						Online Public Notice post date ${postDate} should be on or before commenting start date 
+						${commentingOpenDate}.`);
+					}
+					// Must be at least one day after publish is pushed
+					const dayDiff = DateTimeUtil.diffNow(postDate, DateTimeUtil.TIMEZONE_VANCOUVER, 'day');
+					if (dayDiff < 1) {
+							throw new BadRequestException(`Unable to transition FOM ${entity.id} to ${stateTransition}.  
+							Public Notice Post Date: must be at least one day after publish is pushed.`);
```
Comment (basilv) [inline api/src/app/modules/project/project.service.ts]: I think the terminology of these error messages needs to match what Julius requested for the terminology on the screen "Notice Publish Date" or something like that?

File: api/src/app/modules/project/public-notice.service.ts [hunk 10]
```diff
@@ -104,12 +108,18 @@ export class PublicNoticeService extends DataService<PublicNotice, Repository<Pu
...
       .leftJoinAndSelect("p.forestClient", "forestClient")
       .leftJoinAndSelect("p.district", "district")
-      .andWhere("p.workflow_state_code IN (:...workflowStateCodes)", { workflowStateCodes: [WorkflowStateEnum.COMMENT_OPEN]})
+			// public notices posted earlier than commenting open (in published state with post_date on before today).
+			.where(new Brackets(qb => {
+				qb.where("p.workflow_state_code =:workflowStateCode", {workflowStateCode: WorkflowStateEnum.PUBLISHED})
+				.andWhere("pn.post_date <=:today", {today: DateTimeUtil.nowBC().toDate().toISOString()})
+			}))
+			// public notices for commenting open.
+			.orWhere("p.workflow_state_code IN (:...workflowStateCodes)",
```
Comment (basilv) [inline api/src/app/modules/project/public-notice.service.ts:121]: I'm confused why we check here for COMMENT_OPEN but above we check only for PUBLISHED status. I believe this 'published' is wrong, it should be checking for open. 
(The batch job converts from published to commenting open, only then can they appear in public front-end)

**Review by basilv — CHANGES_REQUESTED** [review]

File: api/src/app/modules/project/public-notice.service.ts (hunk 10 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/public-notice.service.ts:121]: Hi Basil. The "PUBLISHED" status is intended. Our ticket this time is to show public-notice at the "public" side even prior to commenting_open (for the Project). And the only time this happens is when public users go to view on public side on (**today**) is between:  **published pushed date ("PUBLISHED")<= post_date (Public can view) <= commenting_open_date ("COMMENTING_OPEN")**.

So while previously we are checking for just COMMENTING_OPEN that's enough, this time, we also want to include this part "**published pushed date ("PUBLISHED")<= post_date (Public can view)** " for query to find them.

Hope that make sense, it is tested, show and discussed with Julius last week, he confirm that's what he wants to see. If you thinks the query now can be simplify or still wrong, please let me know.

Comment (basilv) [reply api/src/app/modules/project/public-notice.service.ts:121]: Ah, okay that does make sense. It is a shift in how we use the statuses, so we need to remember that going forward. I believe the logic though is if (current date >= post date and status = Published) OR (status = commenting open), then can view the notice. Previously, the batch process ensured that status only goes to commenting open when current date >= commenting open date, so the query doesn't need to check the commenting open date directly. 
So since we don't want to add a new status, we need the current date >= post date when published for the public notices.

So I think it might be okay...

Comment (ianliuwk1019) [reply api/src/app/modules/project/public-notice.service.ts:121]: :)! Thanks Basil.
Yes, "(current date >= post date and status = Published) **OR** (status = commenting open)", that's why there is a "where(**new Brackets(** first condition... **)** ) **.orWhere**( second condition...)"

Comment (basilv) [reply api/src/app/modules/project/public-notice.service.ts:121]: I re-reviewed and am fine with this code.

File: api/src/app/modules/project/project.service.ts (hunk 4 above)
Comment (ianliuwk1019) [reply api/src/app/modules/project/project.service.ts]: Thanks Basil, we will keep this as is.

**Review by basilv — APPROVED** [review]
